### PR TITLE
Revise the `BlockDevice` trait before it's too late

### PR DESCRIPTION
**Pull request by lucassong-mh**

### CODE REVIEW

**Comment by liqinggd** [comment]
LGTM

File: kernel/comps/virtio/src/device/block/device.rs [hunk 1]
```diff
@@ -79,8 +80,12 @@ impl aster_block::BlockDevice for BlockDevice {
         self.queue.enqueue(bio)
     }
 
-    fn max_nr_segments_per_bio(&self) -> usize {
-        self.queue.max_nr_segments_per_bio()
+    fn metadata(&self) -> BlockDeviceMeta {
+        BlockDeviceMeta {
+            max_nr_segments_per_bio: self.queue.max_nr_segments_per_bio(),
+            block_size: BLOCK_SIZE,
```
Comment (tatetian) [inline kernel/comps/virtio/src/device/block/device.rs:86]: Is it possible that `VirtioBlockConfig::blk_size` is not equal to `BLOCK_SIZE`?

Comment (tatetian) [reply kernel/comps/virtio/src/device/block/device.rs]: > What's odd is while using the same qemu configs to boot Linux we got 4096.

How did you got this 4096 number on Linux? You can't print this `VirtioBlockConfig`.

> the blk_size is 512

I think more investigation is needed. If the virtio-blk's `blk_size` is 512, but the kernel thinks the block size is 4096, then this mismatch would cause the kernel to write or read on the wrong blocks. But our file systems on virtio-blk turn out to work correctly. So there must be some misunderstanding.

Comment (lucassong-mh) [reply kernel/comps/virtio/src/device/block/device.rs]: Sure I will get this thing sorted out.

> How did you got this 4096 number on Linux? You can't print this `VirtioBlockConfig`.

I use `blockdev --getbsz /dev/vda` on Linux, which at low level is a `ioctl(BLKBSZSET)`.

Comment (lucassong-mh) [reply kernel/comps/virtio/src/device/block/device.rs:87]: I think I crack the case.

`VirtioBlockConfig::blk_size` will be set to the `logical_block_size` value in qemu cmdline if given. If `logical_block_size` is missed, `blk_size` will be set to the sector size (512 bytes) by default. Check https://github.com/qemu/qemu/blob/master/hw/block/block.c#L144. And `VirtioBlockConfig::capacity` is always the number of 512-bytes sectors.

While the same qemu args to boot Linux, the `virtio_blk_config.blk_size` is also 512 bytes. Check the log below:
image
But `blockdev --getbsz /dev/vda` always returns 4096, which misled me at the first time. (don'y know why)

As for Asterinas, since we donot set `logical_block_size` in qemu args. The `VirtioBlockConfig::blk_size` is always 512 lying there. Although our FSes issue bio requests with 4096-aligned segments, the underlying requests to virtio-blk device are based on sectors, means in the end the segment will be chopped into sectors and be processed, check frontend at https://github.com/asterinas/asterinas/blob/main/kernel/comps/virtio/src/device/block/device.rs#L436 and backend at https://github.com/torvalds/linux/blob/master/drivers/block/virtio_blk.c#L89.

I think we can still issue requests with 4096-aligned block buffers for now, which is the same as the page size, easy to manage, the underlying bio processing will do the sector thing for us. So I still set 4096 in `BlockDeviceMeta` in this PR.

Comment (tatetian) [reply kernel/comps/virtio/src/device/block/device.rs:87]: Thanks for the clarification. To sum up, our `virtio-blk` implementation works with sectors regardless of the real value of `VirtioBlockConfig::blk_size`, which can be specified via QEMU's `logical_block_size` parameter.

This means if QEMU's `logical_block_size` parameter is not 512, then our `virtio-blk` driver will be broken, right? So to improve the robustness of our implementation, I think we should check if the value of `VirtioBlockConfig::blk_size`. If it is not 512, then the driver should panic or return an error.

>  So I still set 4096 in BlockDeviceMeta in this PR.

Why is the use of `BlockDeviceMeta::block_size`? Who cares about it?

Comment (lucassong-mh) [reply kernel/comps/virtio/src/device/block/device.rs:87]: > If it is not 512, then the driver should panic or return an error.

Make sense and done.

> Why is the use of `BlockDeviceMeta::block_size`? Who cares about it?

I will add support for `ioctl(/* any block device */, BLKBSZGET)` and `ioctl(/* any block device */, BLKGETSIZE64)` after this, these basic metadata will be used.

Comment (lucassong-mh) [reply kernel/comps/virtio/src/device/block/device.rs]: More clear out about this "block size" thing after I go deeper:

1. What happened when we want to check block size of a block device in Linux? (via `blockdev --getss/getpbsz/getbsz` /dev/sdX)

In Linux, block device nodes will be added through the `devtmpfs` and `udev` mechanism after boot, we can access them under `/dev` (just `open()` it).
What's behind the device node is mapping to a `S_IFBLK` inode, internally contains a link to [struct block_device](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/blk_types.h#n41) that represents the device. When we issue `ioctl(BLKBSZSET)`, it goes to `static inline unsigned int block_size(struct block_device *bdev) { return 1 << bdev->bd_inode->i_blkbits; }` in `blkdev.h`.
And how did we settle the `i_blkbits`?
Upon a device node is opened via `blkdev_open()`, `set_init_blocksize` in `blkdev.h` is called, which looks like:
```c
static void set_init_blocksize(struct block_device *bdev)
{
  unsigned int bsize = bdev_logical_block_size(bdev);
  loff_t size = i_size_read(bdev->bd_inode);

  while (bsize < PAGE_SIZE) {
    if (size & bsize)
      break;
    bsize <<= 1;
  }
  bdev->bd_inode->i_blkbits = blksize_bits(bsize);
}

static inline unsigned int bdev_logical_block_size(struct block_device *bdev)
{
  return queue_logical_block_size(bdev_get_queue(bdev));
}

static inline unsigned queue_logical_block_size(const struct request_queue *q)
{
  int retval = 512;

  if (q && q->limits.logical_block_size)
    retval = q->limits.logical_block_size;

  return retval;
}
```
So the block size is either the queue's logical block size, or the same value as `PAGE_SIZE` (4096).


2. What are the differences between FSes' block size and devices' block size? In other words, the differences between `fs->superblock->s_blocksize` and `block_device->bd_inode->i_blkbits` in Linux?

When we mount a FS, `sb_min_blocksize()` in `fs/ext2/3/4/super.c` is called, looks like:
```c
/*
 * See what the current blocksize for the device is, and
 * use that as the blocksize.  Otherwise (or if the blocksize
 * is smaller than the default) use the default.
 * This is important for devices that have a hardware
 * sectorsize that is larger than the default.
 */
blocksize = sb_min_blocksize(sb, BLOCK_SIZE);

int sb_min_blocksize(struct super_block *sb, int size)
{
  int minsize = bdev_logical_block_size(sb->s_bdev);
  if (size < minsize)
    size = minsize;
  return sb_set_blocksize(sb, size);
}
```
This means the FS would negotiate the block size with underlying block device.
As for the initial superblock we got from the formatted device, using `mke2fs` for instance, we can set `-b 1024~4096` as configs. `4096` is the default value, since it is the same as the `PAGE_SIZE`, which eases FS for managing cached pages and issue bio requests.


In summary, in Asterinas, as for fs, we use default `mke2fs` to format 4KB-based metadatas, which makes sense since it equals to our page size. As for block device (virtio-blk), we also use the default configuration (logical_block_device=512). This could work fine since we set the right sector range in each bio.

As for `BlockDeviceMeta` of `trait BlockDevice` in this PR, I think adding a `block_size` is basically useless and confusing. So I decide to add a `nr_sectors` only, similar to `block_device->bd_nr_sectors` in Linux. This could not only tell us the capacity of the block device, but also no confusing at all. Sector is a common term in block device and basically equals to 512 anywhere in world (`#define SECTOR_SHIFT 9` in blk_types.h).

Comment (tatetian) [reply kernel/comps/virtio/src/device/block/device.rs]: Excellent write-up! That's the spirit. We should keep getting to the bottom of something before programming it.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
